Modern menu

Malkierian opened the pull request:
This adds a new, more modern-styled menu to use instead of the menubar. The goal was to have it extensible, easier to navigate, and capable of holding a lot more enhancements. I think I accomplished that.

- 3 buttons on the menu header for touch use (when porting to mobile platforms, for instance) for closing the menu, restarting the game, and quitting the program.
- Virtually full size flexibility. Designed around 1280x800, but still functional down to around 400x200.
- All normally extra windows start docked in the menu (for ease of use), but can be popped out to their normal windows.
- The menu itself can be popped out instead of being an overlay, and as such can be docked (could use some refinement, though).
- [Esc] is now the key used to bring it up instead of F1.
- Menu retains its section and subsection selections, meaning you can open and close the menu to toggle an option quickly.
- Menu retains its popped size and position when popped back in.

This also reimplements the gamepad navigation toggle from SoH (which works better in the new menu than on the menubar by several orders of magnitutde), and includes a new enhancement that forces the cursor to be visible even in full screen without the menu or menubar open.

Thanks to @Caladius, this also comes with a search function. It's setup to search widget names and tooltips, and displays each option's location in the rest of the menus.

Note to other devs: the LUS changes will need a bit of tweaking before being made a PR, so that it's not relying on 2ship settings and CVars directly, but is configurable and abstract.

![image](https://github.com/HarbourMasters/2ship2harkinian/assets/774771/7ed6a5cc-1c3e-41a1-a7bf-8e3109224592)
![image](https://github.com/HarbourMasters/2ship2harkinian/assets/774771/3536fdbb-3699-4c0d-a8c7-c5c8a3c98064)
![image](https://github.com/HarbourMasters/2ship2harkinian/assets/774771/1f62190a-e9b6-48da-8954-b1e8c8d17787)

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1979690160.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1979697628.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/1979697989.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Start the framework.

Malkierian committed:
Start header bar.

Malkierian committed:
Header buttons now toggle. Sections show accordingly.
Basic audio section done. Needs subdividing for sane widget widths.

Malkierian committed:
Add fonts and multiple sizes.
Add graphics settings to menu.

Malkierian committed:
Finalized fonts, more sizes.

Malkierian committed:
Modularized the menu structure. Has an errant EndChild somewhere.

Malkierian committed:
Finished first step modularization.
Parameterized menu height.

Malkierian committed:
Some adaptive menu size.

Malkierian committed:
Moved all enhancements over and organized.
Tried to add button option for sidebar entries.

Malkierian committed:
Finished Enhancements Menu.
Sidebar window button still broken (can't reference window variables properly, not drawing in the right place).

Malkierian committed:
Developer Tools header started.
Menu now pops out with a checkbox in Developer Tools.
Menu sizing works for smaller game sizes.
Popout docking disabled.
Port bool added for disallowing popout regardless of cvar.

Malkierian committed:
Better auto-sizing. Columns collapse to 1 below 800 window width, and scroll.

Malkierian committed:
Move inside of HudEditorWindow::DrawElement to DrawContents, and call DrawContents from DrawElement.

Malkierian committed:
Menu resizability complete.
Menu popout complete.
Menu popout position/size is retained when popped in.
Apply HUD editor changes, draw contents in menu if window is closed.
All child sizes properly calculated.
SidebarEntry drawing separated into its own custom element.

Malkierian committed:
Split header entries into separate widget.
Header now dynamically sized and shows scrollbar when necessary.

Malkierian committed:
Modify EventLog to be drawable with the DrawContents setup.
Prevent menu from covering other windows (always in back).

Malkierian committed:
Add reset, close menu, and quit buttons to the header.

Malkierian committed:
Fix reset and quit button placement.

Malkierian committed:
Swap Actor Viewer, Collision Viewer, Save Editor to DrawContents concept.
Change reset icon to FA_ICON_UNDO (single arrow) and quit icon to FA_ICON_POWER_OFF.
Add button to close menu button tooltip, add parentheses to button for reset tooltip.

Malkierian committed:
Custom InputEditorWindow to work with menu.
Fix LUS references in Menu.cpp for Window backend changes.
Implement gamepad navigation and forced cursor visibility checkboxes.
Toggle cursor visibility according to menu status and forced visibility CVar.

Malkierian committed:
clang-format

Malkierian committed:
Swap std for fmt (runner build error).
Ensure cursor visibility is restored if started in full screen with the menu open.

Malkierian committed:
Oops, still had #include <format>

Malkierian committed:
Swap to LUS GetMenu() functionality.
Add option to hide "F1 to open enhancements menu" text at startup if menus aren't open.

Malkierian committed:
Remove last vestiges of pre-LUS-menu functionality.

Malkierian committed:
cleanup

Malkierian committed:
Adapt to GuiElement, GuiWindow changes on the menu support PR.

Malkierian committed:
LUS PR bump.
Adapt to cursor visibility changes in LUS.
Adapt to menu toggle changes in LUS.
Fix input editor not hiding in menu when popped out.

Malkierian committed:
Add new enhancements and cheats to Modern Menu from rebase.

Malkierian committed:
clang

Malkierian committed:
Apply Graphics Menu Lag Fix to Menu structure.

Malkierian committed:
Remove unnecessary window children to simplify controller navigation.

Malkierian committed:
clang

retropieuser commented:
This is looking really good. 

Would it be possible to implement a hotkey to bring the menu up from the gamepad rather than Esc? As you still need a keyboard whereas with this menu we're almost at Big Picture style where this can be played on the tv without a mouse and keyboard required

retropieuser commented:
This is fantastic work, I can finally play things like 2s2h (and hopefully soon soh) on my TV without needing a mouse and keyboard! Thanks for a great job on this

One small tidbit, right now when you have the menu open, pressing gamepad_back only hides the menu while the toggle btn is pressed down, when you let go the menu reappears. To close the menu you have to select the button on the GUI.

Can this be changed so when you press the toggle btn in the menu it closes it please?

Malkierian commented:
It was specifically designed to do that, just like the menubar before it. I'm not experiencing what you're describing. In fact, if I hold the toggle button, it rapidly shows and hides it over and over again.

retropieuser commented:
Ah yes maybe I didn't explain it well. if I hold the toggle button it rapidly turns off and on.

In terms of a user case though and what they're likely to do, just two suggestions:- 

You could disable the toggle button when the menu is visible and re-enabling it when the menu isn't visible? So pressing it once brings up the menu and then you need to select close Menu inside the GUI?

Or you could have the toggle button always able to switch the menu from visible to invisible with a slight wait for presses so it doesn't happen too fast?

Just food for thought anyway. It's an awesome job all the same and thank you for working on it

Malkierian committed:
Add `UpdateWindowBackendObjects()` call to backend combobox.

Malkierian committed:
Rebase to latest develop. Remove redundant window backend stuff that exists now in BenMenuBar.

Caladius committed:
Search Enhancement (#2)

* 90% complete, Commit for review.

* Updated remaining Menu Entries, Reverted Backends and other Window Buttons for now.

Motion Blur copied to Search Menu.

* Did I mention I dislike this clang thing? lol

* Updating MotionBlur options and reverting soem things.

* Integrates .disabled and .disabledTooltip to the constructors. Allows for multiple qualifications to be checked against for disabling widgets and for setting the tooltip accordingly.

* Updated Disabled Reasoning

* Better Disabled State registration system

* Clean up from testing

Malkierian committed:
Expand WidgetOptions to be able to handle string and float values. bool and uint32_t are handled through int32_t.
Numerous renames.
Added several data structures to allow for dynamic updating of CVars for DisableOptions, as well as dynamic evaluation of status with enumerated conditions (less than, greater than, equal to).
Added menuThemeIndex variable to store state of menu theme selection instead of having all the individual calls to CVarGetInteger for it. Updated once per frame.

Malkierian committed:
Changed color vector to map with enum as key and UIWidget::Colors as the value.
More renames.
Moved float slider values into case block.

Malkierian committed:
Add Color_RGB8 and Color_RGBA8 to CVarVariant in preparation for adding color widgets to menu item system.
Not happy it required comparison operator overrides, but it is what it is.

Malkierian committed:
Changed disable conditions to lambdas.

Malkierian committed:
Convert Stats and Collision Viewer to menu item system.
Swap out DisableOption vector for activeDisables and a lambda function for evaluating disabled and hidden. Set the precedent to still use the DisableInfo system.

Malkierian committed:
Add DisableOption for both camera settings being off.
Put Debug Camera in Free Look column, disable common widgets for both cameras, hide widgets specific to either if off.

Malkierian committed:
Moved more widgets to MenuItem system. Some need proper initialization routines because of Ship::Context references.

Malkierian committed:
Convert menu items to nested vector setup.
Remove MenuItem enum.
Add menu item location display to search results.

Malkierian committed:
Partial removal of SidebarEntryIndex enum.

Malkierian committed:
Finish removing SidebarEntryIndex.
Rework drawing code to handle the change back to direct vector access.
Add tooltips to search.

Malkierian committed:
Fix Graphics enhancements section title.

Malkierian committed:
Copy added menu items to modern menu system.

Malkierian committed:
clang

Malkierian committed:
Add some documentation for the menu system components.

Malkierian committed:
Revert change to std::map

Computerfan23 commented:
Hi!

This is really great work and it is really amazing. Will this be able to work better for accessibility using the NVDA screen reader? Like using a text to speech?

Malkierian commented:
It actually should be possible to hook this up directly to the TTS system currently built into SoH when it gets ported over, without too much hassle.

Computerfan23 commented:
That sounds great to hear. It would make it much easier for me to access the enhancements if I can get it readable with tts. Since it would be easier to access using the esc button i believe. Looking forward to when this will be available.

Malkierian committed:
`size_t` because GCC has to be happy

Malkierian committed:
Fixed HUD Editor Widescreen mode not displaying the sliders properly.
Changed all windows to full-section-width display.

Archez commented on mm/2s2h/search/SearchMenu.h:
```diff
@@ -0,0 +1,1377 @@
+                WIDGET_BUTTON,
+                {},
+                [](widgetInfo& info) { CVarSetInteger("gDeveloperTools.FrameAdvanceTick", 1); } },
+              { "Advance (Hold)", "", "Advance frames while the button is held.", WIDGET_BUTTON } } } });
```
Frame advancing "advance (Hold)" is not doing anything/

Malkierian replied:
Done.

Archez commented on mm/2s2h/search/SearchMenu.h:
```diff
@@ -0,0 +1,1377 @@
+                {},
+                [](widgetInfo& info) {
+                    if (CVarGetInteger("gEnhancements.Graphics.MotionBlur.Toggle", 0) == 0) {
+                        R_MOTION_BLUR_ENABLED;
```
All the widget handling around motion blur seems to be incorrect/strange.

`On/Off` is only meant to be used when the setting is Dynamic, and is not supposed to be a CVar as its supposed to reflect the games register values. Similarly, `Strength` when set to dynamic, is also supposed to reflect the games register value, not a CVar.

Additionally, the changes in `MotionBlur.cpp` should be reverted.

Malkierian replied:
Done.

Archez commented on mm/2s2h/search/SearchMenu.h:
```diff
@@ -0,0 +1,1377 @@
+    }
+
+    switch (widget.widgetType) {
+        case WIDGET_CHECKBOX:
```
I think all the cvar widgets should probably have CVAR in the name, and then things like `WIDGET_CHECKBOX` can just be the regular checkbox variant (something the motion blur elements will need)

Malkierian replied:
Done.

Malkierian committed:
Revert changes to MotionBlur.cpp

Malkierian committed:
Add non-CVar variants of checkbox, combobox, and the sliders.
Fix motion blur functionality with new non-CVar widgets.

Malkierian committed:
Add non-CVar variants of checkbox, combobox, and the sliders.
Fix motion blur functionality with those non-CVar widgets.
Add size and labelPosition to widgetOptions, implemented on button and checkbox/combobox/sliders.
Change window popout buttons to inline size type.

Malkierian committed:
Rename SearchMenu.h to SearchableMenuItems.h and move to BenGui folder.

Malkierian committed:
Forgot to change the include XD

Malkierian committed:
Change `modifierFunc` to `preFunc`, updated documentation for that and including `postFunc`.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+                    info.isHidden = disabledMap.at(DISABLE_FOR_MOTION_BLUR_MODE).value != 0 ||
+                                    disabledMap.at(DISABLE_FOR_MOTION_BLUR_OFF).active;
+                },
+                [](widgetInfo& info) { motionBlurStrength = R_MOTION_BLUR_ALPHA; } },
```
I believe the assignment of `motionBlurStrength` should happen as part of the `preFunc`, not the `postFunc`, so that the current frame's value of `R_MOTION_BLUR_ALPHA` is what gets displayed in the slider.

`postFunc` means it is delayed by a frame.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+        case WIDGET_CHECKBOX: {
+            bool* pointer = std::holds_alternative<bool*>(widget.widgetOptions.valuePointer)
+                                ? std::get<bool*>(widget.widgetOptions.valuePointer)
+                                : nullptr;
```
Unfortunately, passing in `nullptr` to `::Checkbox` is unsafe. This will lead to a null pointer dereference and crash.

Instead, we would need to either use a dummy value pointer as the fallback, or bail out of rendering this widget to be safe. We can add a debug `assert()` to at least trip this for local dev.

---

Separately, there are many `std::get()` in this file that are not wrapped with a `std::holds_alternative()` check so things are already inconsistent. Maybe just drop that part?

If we wanted to actually guard against all this, we could just wrap everything with a try/catch for `std::bad_variant_access`

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+        case WIDGET_COMBOBOX: {
+            int32_t* pointer = std::holds_alternative<int32_t*>(widget.widgetOptions.valuePointer)
+                                   ? std::get<int32_t*>(widget.widgetOptions.valuePointer)
+                                   : nullptr;
```
Same thing here.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+                    if (!std::holds_alternative<bool*>(info.widgetOptions.valuePointer) ||
+                        std::get<bool*>(info.widgetOptions.valuePointer) != (bool*)&gPlayState->frameAdvCtx.enabled) {
+                        info.widgetOptions.valuePointer = (bool*)&gPlayState->frameAdvCtx.enabled;
+                    }
```
This logic looks very confusing and is dangerously setting the `valuePointer` with the memory address that reflects the member offset of `frameAdvCtx` against the value of `nullptr` (which means it becomes a non-nullptr value).

I think this should be re-worked as something like:

```c++
if (gPlayState != nullptr) {
    info.widgetOptions.valuePointer = (bool*)&gPlayState->frameAdvCtx.enabled
} else {
    info.widgetOptions.valuePointer = (bool*)nullptr // or dummy pointer
}
```

Additionally, we probably don't want the initializer to assign with `gPlayState` as that is going to be null during init and will experience the same behavior.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+    CVarVariant max;
+    CVarVariant defaultVariant;
+    std::unordered_map<int32_t, const char*> comboBoxOptions;
+    std::variant<bool*, int32_t*, float*> valuePointer;
```
[non-blocking]

The default initialized value for `valuePointer` ends up as `(bool*)nullptr` and the variant type as `bool*`.

We might want to consider allowing `nullptr_t` as a type so that we can explicitly deal with `nullptr` values.

Marking as non-blocking for this merge, but probably something to re-visit for when upstreaming to LUS.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+            }
+            break;
+        case WIDGET_SLIDER_INT:
+            if (UIWidgets::SliderInt(widget.widgetName, std::get<int32_t*>(widget.widgetOptions.valuePointer),
```
Same thing above about `nullptr` is bad for Sliders cause of dereference (and below for SliderFloat)

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+                                ImGui::PushStyleColor(ImGuiCol_Text, UIWidgets::Colors::Gray);
+                                std::string origin =
+                                    fmt::format("  ({} -> {}, Clmn {})", menuLabel, sidebar.label, column);
+                                ImGui::Text(origin.c_str());
```
`warning: format string is not a string literal (potentially insecure) [-Wformat-security]`

```suggestion
                                ImGui::Text("%s", origin.c_str());
```

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+            }
+            break;
+        case WIDGET_WINDOW_BUTTON: {
+            if (widget.widgetOptions.windowName == "") {
```
`result of comparison against a string literal is unspecified (use an explicit string comparison function instead) [-Wstring-compare]`

`strlen` or this to be safer

```suggestion
            if (widget.widgetOptions.windowName == nullptr || widget.widgetOptions.windowName[0] == '\0') {
```

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+    MOTION_BLUR_DYNAMIC,
+    MOTION_BLUR_ALWAYS_OFF,
+    MOTION_BLUR_ALWAYS_ON,
+};
```
Missing typedef name
```suggestion
} MotionBlurOption;
```

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1514 @@
+    DEBUG_LOG_ERROR,
+    DEBUG_LOG_CRITICAL,
+    DEBUG_LOG_OFF,
+};
```
Missing typedef name

```suggestion
} DebugLogOption;
```

Malkierian replied:
To be fair, these typedef names weren't there when they were originally added, so it's not my fault, but I'll add them XD.

Archez reviewed (commented):
Ok just a few more things (mainly around the new changes and some warning fixes)

Malkierian committed:
Modify code to require variable for non-CVar widget initialization when necessary. Check for `valuePointer` being nullptr, and preventing widget drawing with a debug assert and error log entry.
Remove all `has_variant` calls and wrap entire widget type switch in `SearchMenuGetItem` with a try/catch on `std::bad_variant_access` with another error log and debug assert.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1539 @@
+            widget.postFunc(widget);
+        }
+    } catch (const std::bad_variant_access& e) {
+        SPDLOG_ERROR(e.what());
```
```suggestion
        SPDLOG_ERROR("Failed to draw menu item \"{}\" due to: {}", widget.widgetName, e.what());
```

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1539 @@
+                    }
+                },
+                [](widgetInfo& info) {
+                    if (disabledMap.at(DISABLE_FOR_NOT_DIRECTX).active) {
```
Looks like in last commit  removed the hide for the non-DX match refresh rate checkbox.

Probably should add back the other hide so there aren't two match refresh options on DX.

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1539 @@
+           return Ship::Context::GetInstance()->GetWindow()->GetWindowBackend() !=
+                  Ship::WindowBackend::FAST3D_DXGI_DX11;
+       },
+        "Multi-viewports not supported" } },
```
Copy paste of previous disabled message. Should instead be something like "Only available on DirectX"

Archez commented on mm/2s2h/BenGui/SearchableMenuItems.h:
```diff
@@ -0,0 +1,1539 @@
+
+              { "Current FPS: %d",
+                "gInterpolationFPS",
+                tooltip.c_str(),
```
The pointer value produced by `std::string::c_tr()` is only valid while the parent string is alive. Because `topic` is a function scope variable, it gets cleaned up once `AddSettings()` finishes, and then the resulting text displayed in the tool tip is garbage.

You will either need to keep this variable alive by moving it outside of the function scope, or marking it as `static`.

Malkierian replied:
Now that I look at it, there was no reason for it to be std::string to begin with, changing `tooltip` to `const char*` seems to have resolved the issue.

Malkierian committed:
More review changes.

Archez reviewed (approved):
Thanks for your patience with the reviewing, I feel good with where this PR is at now. Once in, I'm sure we will get good feedback on further improvements we can make.

garrettjoecox commented:
~Idk what setting I had set, but I couldn't interact with the new menu until I deleted my JSON/Imgui.init files. Probably want to figure out what setting that is before we have to deal with it in support.~ Unable to reproduce unfortunately

Blocking:
- There seems to be some weirdness regarding whether or not the menu shows up when you press your controller's dedicated button. When I open up the game, my controller's `menu` button doesn't do anything. If I then click ESC on my keyboard to open the menu, then "grab focus" with my controller by moving the dpad, and then navigate to and close the menu with the "Close Menu" button in the UI, my controller seems to retain focus, and then the `menu` button on my controller works to re-open the menu. Additionally, regardless of if the menu has focus or not, when it is open and I try to close it using my controller's `menu` button, it closes and re-opens.

Not blocking:
- The icons within buttons are not vertically centered. I think this might have been an issue before, but with the larger font size it seems more exaggerated.
- Volume sliders on audio page are wrong format
- Some things in the wrong section or missing, eg 2 options under "time savers > dialog" don't belong, and warp point developer tool is missing
- I know this is obvious, but the navigation buttons & search should use our themed widgets
- We should just get rid of the "Press f1" popup all together, pressing your menu button or escape seems intuitive enough
- When the search page is opened, the input should be auto focused
- Menu.cpp/.h should both be under the BenGui/ folder
- Not sure how I feel yet about the pattern in which we define menu items. Probably won't know how I feel until I have to work with it for a while.

What is the plan for the old Navbar? It's going to get annoying real quick to have to add UI stuff in two places

Malkierian committed:
Implement `isPercentage`, `format`, and `showButtons` for sliders.
Pull values of those over from menubar.
Add documentation on `widgetOptions` for those and others added for previous review changes.
Move Fast Magic Arrow and Instant Fin Recall to Gameplay section.

Malkierian commented:
For the major problem, it's going to require some investigation. As usual, what I was developing on, WASAPI, has no issues with the way things are. Those issues only show up with SDL as the window manager, and the fact that they're handled differently between the two annoys me to no end, but that's where they end up manifesting.

For the non-blocking stuff:
1) I'm not even sure how to make that center. The problem seems to be that FA renders the icons at the text line floor, and I don't know if there's a way to change that.
2) Easily enough handled.
3) Yeah, the dialog things were supposed to go in Gameplay, I think. Not sure how that happened. And I knew I'd be missing some things XD
4) The nav buttons are easy enough, but I don't think the search widget has a themed widget variant
5) I can see the the prompt skip default to on, once we decide to yeet the navbar. The only reason it isn't is because I didn't know whether to yeet that in this PR or not. I figured people would want to have the option of it while getting used to the new menu.
6) I'm really not sure how to give the search bar focus only on first draw when changing to the search tab...
7) Also easily done
8) Any suggestions for improvement in that flow will always be welcome. It's already gone through 4 or 5 iterations between what I originally did and what Cal did for the search menu, and then what I did to further extend what he did. This is very much still WIP, it's just functional so that more people can test it.

Malkierian committed:
Modify header and sidebar "buttons" to match theming of regular buttons.

Malkierian committed:
Add precision clamping to SliderFloat value. At least 1.0 seems to be consistent at "100%".
Change IR slider to be percentage.

Malkierian committed:
Mention weirdness with SDL and pad menu button in controller nav tooltip.

Malkierian committed:
Move Menu files to 2s2h/BenGui.

Malkierian committed:
Add Warp Points to Developer Tools.
Make "Press F1" prompt show only when a menu isn't registered.

Malkierian committed:
clang

Malkierian committed:
Make search input take focus when search menu is open.

Malkierian committed:
Forgot include.

garrettjoecox reviewed (approved):
Let's make an issue to track and detail the SDL thing

garrettjoecox committed:
Global search (#3)

Malkierian committed:
Restore Sidebar search entry, but lock it behind a checkbox in Settings. Dynamically adds or removes the sidebar entry, and header search is toggled inversely.

Malkierian commented:
> Let's make an issue to track and detail the SDL thing

Handled in https://github.com/Kenix3/libultraship/pull/677

Malkierian committed:
Make menu toggle its visibility when using the button in the menu to close. Couldn't handle Alt+F4 without a hook I didn't want to add.
Remove unneeded commented code.
